### PR TITLE
New model PR needs green (slow tests) CI

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

For a new model PR, let's trigger and run the slow tests of that model.

Question: if we want to trigger the slow tests for each commit in a PR, or we prefer to trigger it toward the end of the PR (i.e. almost ready to be merged to main)? In the later case, we can use the commit message (i.e. having a prefix like `[run_slow]`).

### CODE REVIEW

File: .github/workflows/self-new-model-pr-caller.yml
Comment (ydshieh) [inline .github/workflows/self-new-model-pr-caller.yml:4]: we could probably also trigger this on `merge to main event`, but the point of this PR (so far) is to make sure the new model PR has green tests for that model.

File: .github/workflows/self-new-model-pr-caller.yml
Comment (ydshieh) [inline .github/workflows/self-new-model-pr-caller.yml:123]: We want to block the merge button if a new model PR has any failing  slow test. But since the previous job `run_test` may have a `skipped` status (if there is no new model), we still need a job (i.e. this one) to indicate it as `successful` so we can still show the merge button in this case.

File: .github/workflows/self-new-model-pr-caller.yml
Comment (ArthurZucker) [inline .github/workflows/self-new-model-pr-caller.yml:41]:
```suggestion
      if: needs.check_for_new_model.outputs.new_model != '' && contains(github.actor, 'transformers')
```

Comment (ArthurZucker) [reply .github/workflows/self-new-model-pr-caller.yml:41]: let's make sure only HF people can trigger this for safety reasons

File: .github/workflows/self-new-model-pr-caller.yml
Comment (ArthurZucker) [inline .github/workflows/self-new-model-pr-caller.yml:71]: not sure I understand that? We are already in transformers branch no?

File: .github/workflows/self-new-model-pr-caller.yml
Comment (ArthurZucker) [inline .github/workflows/self-new-model-pr-caller.yml:19]: Let's only do torch for now no?

File: .github/workflows/self-new-model-pr-caller.yml
Comment (ArthurZucker) [inline .github/workflows/self-new-model-pr-caller.yml:92]:
```suggestion
        run: python3 -m pytest -v -rs --make-reports=${{ matrix.machine_type }}_tests_gpu_${{ matrix.folders }} tests/${{ matrix.folders }}
```
let's make sure we know why some tests are skipped

**Review by ArthurZucker — COMMENTED** [review]
🤗

File: .github/workflows/self-new-model-pr-caller.yml
Comment (ydshieh) [reply .github/workflows/self-new-model-pr-caller.yml:71]: Here we use docker image, and we don't use `actions/checkout`, so it will be at the `main` branch (at the commit when the docker image is built).

We can definite try using `actions/checkout`, but all the paths in all workflow files have to be updated for which we don't gain much.

(and even if we use `actions/checkout`, we still need to run `pip install -e .`)

File: .github/workflows/self-new-model-pr-caller.yml
Comment (amyeroberts) [reply .github/workflows/self-new-model-pr-caller.yml:41]: This is great! Nice to have this way to control instead of having to request specific commit messages or running all the time

File: .github/workflows/self-new-model-pr-caller.yml
Comment (amyeroberts) [reply .github/workflows/self-new-model-pr-caller.yml:19]: I'd like to keep `RUN_PT_TF_CROSS_TESTS=1` - if there's any changes made to an important with a TF port which require slow tests to be run, we want to make sure the models are still equivalent

File: .github/workflows/self-new-model-pr-caller.yml
Comment (ydshieh) [reply .github/workflows/self-new-model-pr-caller.yml:41]: @amyeroberts [My above comment](https://github.com/huggingface/transformers/pull/30341#discussion_r1574831835) only works for outside contributors. For internal contributors, the workflow won't need to be approved to be run - hence every commit will trigger the workflow (for which I think it's not great, but probably acceptable)

Comment (ArthurZucker) [reply .github/workflows/self-new-model-pr-caller.yml:41]: Mmmm I am not in favor of every comming triggering the workflow. When you add the model you don't want to do this!

File: .github/workflows/self-new-model-pr-caller.yml
Comment (ArthurZucker) [reply .github/workflows/self-new-model-pr-caller.yml:71]: ah alright the same old issue!

File: utils/check_if_new_model_added.py
Comment (amyeroberts) [inline utils/check_if_new_model_added.py:54]: Do we want to just constrain to adding new models? If someone is touching modeling code for e.g. llama, it's important those slow tests are run too. 

Maybe just asking people to run with `[run-slow]` in the commit message it enough for these cases, but it would for them to be picked up

**Review by amyeroberts — APPROVED** [review]
Thanks for adding! 

As @ArthurZucker says - running for all commits for maintainers isn't ideal. If necessary, having `[run-slow]` flag in the commit message seems OK, but would be great to have this as something which is necessary for merge but we can also control from within the github web environment

File: utils/check_if_new_model_added.py
Comment (ydshieh) [reply utils/check_if_new_model_added.py:54]: The problem of other cases is: they might contain multiple files touched and take too long to run. For example, a pull could actually trigger the whole CI. In fact, a new model PR would trigger the whole CI too  as it touch `transformers.__init__`.

The new workflow file added in this PR has a modification:  it runs a single model (as the new model PR has a well-defined target model to run: it's the new model to be test against).

In general cases, if we trigger all modified modeling files - it might take too long (and possibly from multiple PRs)

Comment (amyeroberts) [reply utils/check_if_new_model_added.py:54]: OK - can we still enable running with a `run_slow` commit flag in a follow-up, such that we can request if and when it's needed

Comment (ydshieh) [reply utils/check_if_new_model_added.py:54]: > OK - can we still enable running with a run_slow commit flag in a follow-up, such that we can request if and when it's needed

For this, do  you mean run all tests (even slow ones) that are detected by the test fetcher?

**Comment by ydshieh** [comment]
Convert to draft to check if I can using some github action events to control the CI trigger more precisely.

File: .github/workflows/self-new-model-pr-caller.yml
Comment (ydshieh) [inline .github/workflows/self-new-model-pr-caller.yml:26]: After looking into the GitHub Actions doc and online discussion, I decide to add this condition.

**Only** after the PR has been labeled with `single-model-run-slow`, the subsequent `pull_request` events (like push more commits to the PR) will trigger the workflow.

Note that, the **push** after being labeled by `single-model-run-slow` is still necessary to trigger the CI.
I avoid to trigger the CI by using `issue_comment` (as this is way too frequent, despite we can filter it out later).

So please ask the contributors to push an empty commit after add this label. For external contributors' PR, we still need to click `Approve and run` button.


We can explore the `pull_request: labeled` event in the future.

Comment (ydshieh) [reply .github/workflows/self-new-model-pr-caller.yml:26]: > add the label single-model-run-slow
This is the requirement, but it won't actually trigger the CI. The contributor has to push a commit (even empty) to trigger it. 

But we can explore to use another event **together** where when we add a label (`pull_request: labeled`). In that event, the CI will be triggered (if the added label is `single-model-run-slow`).

Why we don't just use the above: because the PR might need multiple CI runs before it is ready to be merged. And we don't want the reviewers to add label / remove it / re-add label  to achieves this.

**Comment by ydshieh** [comment]
Sorry I have some trouble to make this PR work with the `merge button blocked`. I need to figure out a better condition.

**Pull request merged**
